asset-transfer-secured-agreement/chaincode-go/asset_transfer.go

```diff
@@ -273,211 +252,228 @@ func (s *SmartContract) TransferAsset(ctx contractapi.TransactionContextInterfac
+		return fmt.Errorf("failed to create composite key: %v", err)
 	}
 	sellerPriceHash, err := ctx.GetStub().GetPrivateDataHash(collectionSeller, assetForSaleKey)
 	if err != nil {
-		return fmt.Errorf("failed to get seller price hash: %s", err.Error())
+		return fmt.Errorf("failed to get seller price hash: %v", err)
 	}
 	if sellerPriceHash == nil {
 		return fmt.Errorf("seller price for %s does not exist", asset.ID)
 	}
 
-	// get buyer bid price
+	// Get buyers bid price
 	collectionBuyer := buildCollectionName(buyerOrgID)
 	assetBidKey, err := ctx.GetStub().CreateCompositeKey(typeAssetBid, []string{asset.ID})
 	if err != nil {
-		return fmt.Errorf("failed to create composite key: %s", err.Error())
+		return fmt.Errorf("failed to create composite key: %v", err)
 	}
 	buyerPriceHash, err := ctx.GetStub().GetPrivateDataHash(collectionBuyer, assetBidKey)
 	if err != nil {
-		return fmt.Errorf("failed to get buyer price hash: %s", err.Error())
+		return fmt.Errorf("failed to get buyer price hash: %v", err)
 	}
 	if buyerPriceHash == nil {
 		return fmt.Errorf("buyer price for %s does not exist", asset.ID)
 	}
 
-	// get sha256 hash of passed price
 	hash = sha256.New()
 	hash.Write(priceJSON)
 	calculatedPriceHash := hash.Sum(nil)
 
-	// verify that the hash of the passed price matches the on-chain seller price hash
+	// Verify that the hash of the passed price matches the on-chain sellers price hash
 	if !bytes.Equal(calculatedPriceHash, sellerPriceHash) {
-		return fmt.Errorf("hash %x for passed price JSON %s does not match on-chain hash %x, seller hasn't agreed to the passed trade id and price", calculatedPriceHash, priceJSON, sellerPriceHash)
+		return fmt.Errorf("hash %x for passed price JSON %s does not match on-chain hash %x, seller hasn't agreed to the passed trade id and price",
+			calculatedPriceHash,
+			priceJSON,
+			sellerPriceHash,
+		)
 	}
 
-	// verify that the hash of the passed price matches the on-chain buyer price hash
+	// Verify that the hash of the passed price matches the on-chain buyer price hash
 	if !bytes.Equal(calculatedPriceHash, buyerPriceHash) {
-		return fmt.Errorf("hash %x for passed price JSON %s does not match on-chain hash %x, buyer hasn't agreed to the passed trade id and price", calculatedPriceHash, priceJSON, buyerPriceHash)
+		return fmt.Errorf("hash %x for passed price JSON %s does not match on-chain hash %x, buyer hasn't agreed to the passed trade id and price",
+			calculatedPriceHash,
+			priceJSON,
+			buyerPriceHash,
+		)
 	}
 
-	// since all checks passed, return without an error
 	return nil
 }
 
-// transferAssetState makes the public and private state updates for the transferred asset
+// transferAssetState performs the public and private state updates for the transferred asset
 func transferAssetState(ctx contractapi.TransactionContextInterface, asset *Asset, immutablePropertiesJSON []byte, clientOrgID string, buyerOrgID string, price int) error {
```

Comment:
This function seems very dangerous. It has many error paths, and after the initial PutState, any one of those error paths could leave the asset in a state where multiple orgs own it, multiple orgs can modify it, among other things. There is no logic to handle a recovery in the even an error occurs.

Reply:
> This function is very dangerous. It has many error paths, and after the initial PutState, any one of those error paths could leave the asset in a state where multiple orgs own it, multiple orgs can modify it, among other things. There is no logic to handle a recovery in the event an error occurs.

@btl5037 If it errors at any point, the error is returned to the client rather than a partial rwset. So the writes will be atomic, either they all succeed, or there is an error returned.  Maybe I'm not understanding the concern, can you elaborate?